frontend/z3_axioms.py

```diff
@@ -1,203 +1,404 @@
+    """
     return [
-        Implies(new != zNone, subtype(new, result))
+        Implies(new != types.none, types.subtype(new, result))
     ]
 
 
-def control_flow(body, orelse, result):
+def control_flow(then, orelse, result, types):
+    """Constraints for control-flow blocks (if/else, while, for)"""
     # TODO numeric casting
     return [
-        Implies(orelse == zNone, result == body),
-        Implies(orelse != zNone, And(
-            subtype(body, result),
-            subtype(orelse, result)
+        Implies(orelse == types.none, result == then),
+        Implies(orelse != types.none, And(
+            types.subtype(then, result),
+            types.subtype(orelse, result)
         ))
     ]
 
 
-def for_loop(iterable, target):
+def for_loop(iterable, target, types):
+    """Constraints for for-loop iterable and iteration target"""
+    x = Const("x", types.type_sort)
     return [
         Or(
-            iterable == List(target),
-            iterable == Set(target),
-            Exists([x], iterable == Dict(target, x), patterns=[Dict(target, x)]),
-            And(iterable == String, target == String),
-            And(iterable == Bytes, target == Bytes)
+            iterable == types.list(target),
+            iterable == types.set(target),
+            Exists([x], iterable == types.dict(target, x), patterns=[types.dict(target, x)]),
+            And(iterable == types.string, target == types.string),
+            And(iterable == types.bytes, target == types.bytes)
         )
     ]
 
 
-def try_except(body, orelse, final, result):
+def try_except(then, orelse, final, result, types):
+    """Constraints for try/except block"""
     return [
-        subtype(body, result),
-        subtype(orelse, result),
-        subtype(final, result)
+        types.subtype(then, result),
+        types.subtype(orelse, result),
+        types.subtype(final, result)
     ]
+
+
+def instance_axioms(called, args, result, types):
+    """Constraints for class instantiation"""
+
+    if len(args) + 1 >= len(types.funcs):  # Instantiating a class with more number of args than the max possible number
+        return []
+
+    # Assert with __init__ function of all classes in the program
+    axioms = []
+    for t in types.all_types:
+        instance = getattr(types.type_sort, "instance")(types.all_types[t])
+        init_func = types.attributes[t]["__init__"]
+        axioms.append(
+            And(called == types.all_types[t],
+                result == instance,
+                init_func == types.funcs[len(args) + 1]((instance,) + args + (types.none,))))
+
+    return axioms
+
+
+def call(called, args, result, types):
+    """Constraints for calls
+    
+    Cases:
+        - Function call
+        - Class instantiation
+    """
+    return [
+        Or(
+            [called == types.funcs[len(args)](args + (result,))] + instance_axioms(called, args, result, types)
```

Comment:
Because every `Call` node is either a function call or a class instantiation, and we may never statically know which is the case. So for every call, we assert on both scenarios and let Z3 resolve the valid one.